src/evaluate/evaluator/token_classification.py

```diff
@@ -0,0 +1,277 @@
+# Copyright 2022 The HuggingFace Datasets Authors and the TensorFlow Datasets Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any, Callable, Dict, List, Optional, Tuple, Union
+
+from datasets import ClassLabel, Dataset, Sequence
+from typing_extensions import Literal
+
+from .base import Evaluator
+
+
+class TokenClassificationEvaluator(Evaluator):
+    """
+    Token classification evaluator.
+
+    This token classification evaluator can currently be loaded from [`evaluator`] using the default task name
+    `token-classification`.
+
+    Methods in this class assume a data format compatible with the [`TokenClassificationPipeline`].
+    """
+
+    PIPELINE_KWARGS = {"ignore_labels": []}
+
+    def __init__(self, task="token-classification", default_metric_name=None):
+        super().__init__(task, default_metric_name=default_metric_name)
+
+    def predictions_processor(self, predictions: List[List[Dict]], words: List[List[str]], join_by: str):
+        preds = []
+
+        for i, prediction in enumerate(predictions):
+            pred_processed = []
+            label_offsets = self.words_to_offsets(words[i], join_by)
+
+            token_index = 0
+            for label_offset in label_offsets:
+                while prediction[token_index]["start"] < label_offset[0]:
+                    token_index += 1
+
+                if prediction[token_index]["start"] > label_offset[0]:  # bad indexing
+                    pred_processed.append("O")
```

Comment:
since the tokenizer usually at least splits on whitespaces the pipeline shouldn't predict a start anywhere inside a word unless it's the start of the second token which would also be wrong, right?

I am not very familiar with POS but looking as seqeval and the transformers code there doesn't seem to be a special treatment. As long we have `Trainer` parity I would say we are reasonably in the clear. Wdyt?

Reply:
> the tokenizer usually at least splits on whitespaces

I think this is not true. See e.g. https://huggingface.co/abhishek/autonlp-japanese-sentiment-59363/blob/main/vocab.txt#L23394

We may have a dataset as https://huggingface.co/datasets/msra_ner , where each ideogram is split individually, we will join with `""`, but then we may have a tokenizer that will tokenize a single token two ideograms! So it is just to be safe there, although I think we should not be too concerned, this is a very edge case.

> As long we have Trainer parity I would say we are reasonably in the clear. Wdyt?

Fully agree. There is no `run_pos.py` in transformers example though. So I will just add a test in `test_evaluator.py`.